cabal-install/main/Main.hs

```diff
@@ -907,10 +908,19 @@ unpackAction getFlags extraArgs globalFlags = do
   getAction getFlags extraArgs globalFlags
 
 initAction :: InitFlags -> [String] -> Action
-initAction initFlags extraArgs globalFlags
-    | not (null extraArgs) =
-      die' verbosity $ "'init' doesn't take any extra arguments: " ++ unwords extraArgs
-    | otherwise = do
+initAction initFlags extraArgs globalFlags = do
+  -- it takes the first value within extraArgs (if there's one)
+  -- and uses it as the root directory for the new project
+  case extraArgs of
+    projectDir : [] -> do
+      createDirectoryIfMissing True projectDir
+      withCurrentDirectory projectDir initAction'
+    _ : _ -> die' verbosity $
+      "'cabal init' only receives a single, optional," ++
+      "extra argument for the project root directory"
+    [] -> initAction'
```

Comment:
A few small suggestions, reorder cases, use createDirectoryIfMissingVerbose, formatting on the message to be more consistent with the rest of the file. Sorry for being a bit nitpicky.
```suggestion
  case extraArgs of
    [] -> initAction'
    [projectDir] -> do
      createDirectoryIfMissingVerbose verbosity True projectDir
      withCurrentDirectory projectDir initAction'
    _ -> die' verbosity $ "'init' only takes a single, optional,"
                       ++ "extra argument for the project root directory"
```